# Probability
# Introduction to Probability
# Step 1 – Understand the Concept
# Definition
Probability is a measure of how likely an event is to happen.

The probability value is always between 0 and 1.

0 means the event is impossible.
1 means the event is certain.
A value between 0 and 1 represents different levels of likelihood.
Formula
Probability of an event:

P(A) = Number of favorable outcomes / Total number of possible outcomes

# Step 2 – Demonstrate the Concept
Numerical Example
Suppose we roll a fair six-sided die.

Possible outcomes:

1, 2, 3, 4, 5, 6

Probability of getting a 4:

P(4) = 1 / 6

P(4) = 0.1667

So, there is approximately a 16.67% probability of getting a 4.

# Real-World Example
Weather forecasting uses probability.

For example, if the probability of rain is 70%, rain is considered more likely than if the probability were 20%.

# Business Example
An e-commerce company may estimate the probability that a customer will purchase a product after viewing it.

# AI/ML Example
Classification models often produce probabilities for different classes.

For example, a model may predict:

Cat = 0.80 Dog = 0.20

This means the model estimates an 80% probability for the cat class.

How does this demonstrate the concept?
Probability converts uncertainty about an event into a numerical value between 0 and 1.

# Step 3 – Implement the Concept

In [2]:

favorable_outcomes = 1
total_outcomes = 7

probability = favorable_outcomes / total_outcomes

print("Probability of getting a 4:", probability)
print("Probability in percentage:", probability * 100, "%")

Probability of getting a 4: 0.14285714285714285
Probability in percentage: 14.285714285714285 %


# What Does the Program Do?
The program calculates the probability of getting a 4 when rolling a fair six-sided die.

# What Statistical Concept Does It Demonstrate?
It demonstrates basic probability.

# How Do We Interpret the Output?
The probability is approximately 0.1667, or 16.67%.

# Where Is It Used in AI/ML?
Probability is used in classification, prediction, uncertainty estimation, and probabilistic models.

# Events
# Step 1 – Understand the Concept
# Definition
An event is a specific outcome or a group of outcomes that we are interested in from a sample space.

For example, when rolling a die:

Getting a 4 is an event.
Getting an even number is also an event.
An event can contain one or more possible outcomes.

# Step 2 – Demonstrate the Concept
# Numerical Example
When rolling a die:

Sample space:

{1, 2, 3, 4, 5, 6}

Event A = getting an even number.

A = {2, 4, 6}

There are 3 favorable outcomes out of 6.

P(A) = 3 / 6 = 0.5

# Real-World Example
For a customer visiting an online store, an event could be:

"Customer makes a purchase."

# Business Example
A business can define an event as:

"Customer purchases a product within 7 days."

The company can then calculate the probability of this event.

# AI/ML Example
In classification, an event can represent a particular class prediction, such as:

"Customer will churn."

# How does this demonstrate the concept?
An event identifies the specific outcome or group of outcomes we want to analyze.

# Step 3 – Implement the Concept

In [3]:
outcomes = [1, 2, 3, 4, 5, 6]

even_numbers = [x for x in outcomes if x % 2 == 0]

probability = len(even_numbers) / len(outcomes)

print("Even numbers:", even_numbers)
print("Probability:", probability)

Even numbers: [2, 4, 6]
Probability: 0.5


# What Does the Program Do?
The program identifies the even outcomes from a die and calculates their probability.

# What Statistical Concept Does It Demonstrate?
It demonstrates how an event is represented using a set of favorable outcomes.

# How Do We Interpret the Output?
There are three even outcomes: 2, 4, and 6.

Therefore:

P(Even) = 3 / 6 = 0.5

# Where Is It Used in AI/ML?
Events can represent outcomes such as fraud detection, customer churn, or a predicted class.

# Sample Space
# Step 1 – Understand the Concept
# Definition
Sample space is the complete set of all possible outcomes of a random experiment.

It is usually represented by S.

# Example
When tossing a coin:

S = {Head, Tail}

When rolling a die:

S = {1, 2, 3, 4, 5, 6}

# Step 2 – Demonstrate the Concept
Numerical Example
If a coin is tossed once:

Sample space:

S = {H, T}

There are two possible outcomes.

Real-World Example
Suppose a company wants to predict whether a customer will purchase a product.

The possible outcomes can be:

S = {Purchase, No Purchase}

# Business Example
A bank may define possible loan application outcomes as:

{Approved, Rejected}

# AI/ML Example
For a binary classification problem, the sample space can contain two classes such as:

{Spam, Not Spam}

# How does this demonstrate the concept?
The sample space contains every possible outcome that can occur in the experiment.

# Step 3 – Implement the Concept

In [4]:

sample_space = ["Head", "Tail"]

print("Sample Space:", sample_space)
print("Number of possible outcomes:", len(sample_space))

Sample Space: ['Head', 'Tail']
Number of possible outcomes: 2


# What Does the Program Do?
The program creates the sample space for one coin toss.

# What Statistical Concept Does It Demonstrate?
It demonstrates the complete set of possible outcomes.

# How Do We Interpret the Output?
The coin has two possible outcomes: Head and Tail.

# Where Is It Used in AI/ML?
Sample spaces help define possible outcomes in probabilistic models and classification problems.

# Independent Events
# Step 1 – Understand the Concept
# Definition
Two events are independent when the occurrence of one event does not affect the probability of the other event.

Formula
P(A and B) = P(A) × P(B)

# Step 2 – Demonstrate the Concept
Numerical Example
Suppose we toss a coin and roll a die.

Event A = getting Head.

P(A) = 1/2

Event B = getting 6.

P(B) = 1/6

The coin result does not affect the die result.

Therefore:

P(A and B) = 1/2 × 1/6

= 1/12

# Real-World Example
If two separate fair dice are rolled, the result of the first die does not affect the result of the second die.

# Business Example
Suppose two unrelated customers independently make purchase decisions.

One customer's decision does not directly determine the other customer's decision.

# AI/ML Example
Naive Bayes classifiers make a conditional independence assumption between features given the class.

# How does this demonstrate the concept?
The probability of both events occurring is found by multiplying their individual probabilities because the events are independent.

# Step 3 – Implement the Concept

In [5]:
probability_a = 1 / 3
probability_b = 1 / 6

probability_both = probability_a * probability_b

print("P(A):", probability_a)
print("P(B):", probability_b)
print("P(A and B):", probability_both)

P(A): 0.3333333333333333
P(B): 0.16666666666666666
P(A and B): 0.05555555555555555


# What Does the Program Do?
The program calculates the probability of two independent events occurring together.

# What Statistical Concept Does It Demonstrate?
It demonstrates the multiplication rule for independent events.

# How Do We Interpret the Output?
The probability of getting Head and then rolling a 6 is:

1/12 ≈ 0.0833

So the probability is approximately 8.33%.

# Where Is It Used in AI/ML?
Independence assumptions are important in probabilistic models such as Naive Bayes.

# Dependent Events
# Step 1 – Understand the Concept
# Definition
Two events are dependent when the occurrence of one event changes the probability of the other event.

Formula
P(A and B) = P(A) × P(B|A)

Here P(B|A) means the probability of B occurring after A has already occurred.

# Step 2 – Demonstrate the Concept
# Numerical Example
A bag contains:

3 red balls 2 blue balls

Total = 5 balls.

Event A = selecting a red ball first.

P(A) = 3/5

After selecting one red ball, the bag contains:

2 red balls 2 blue balls

Now:

P(B|A) = 2/4

Therefore:

P(A and B) = 3/5 × 2/4

= 6/20

= 0.30

# Real-World Example
Suppose a company selects employees for training without replacing selected employees.

After one employee is selected, the number of remaining employees changes.

Therefore, the probability of selecting the next employee changes.

# Business Example
When selecting products from a limited inventory without replacement, the first selection changes the available inventory for the second selection.

# AI/ML Example
Dependent events are important in sequential data and probabilistic models where previous observations can influence future observations.

# How does this demonstrate the concept?
The first selection changes the situation for the second selection, so the events are dependent.

# Step 3 – Implement the Concept

In [6]:
probability_first_red = 4 / 6
probability_second_red = 3/ 7

probability_both = probability_first_red * probability_second_red

print("Probability of first red:", probability_first_red)
print("Probability of second red given first red:", probability_second_red)
print("Probability of both red:", probability_both)

Probability of first red: 0.6666666666666666
Probability of second red given first red: 0.42857142857142855
Probability of both red: 0.2857142857142857


# What Does the Program Do?
The program calculates the probability of selecting two red balls without replacement.

# What Statistical Concept Does It Demonstrate?
It demonstrates dependent events.

# How Do We Interpret the Output?
The probability of selecting two red balls is 0.30 or 30%.

The second probability changes because one red ball was already removed.

# Where Is It Used in AI/ML?
Dependency between observations is important in sequential models, time-series analysis, and probabilistic modeling.

# Conditional Probability
# Step 1 – Understand the Concept
# Definition
Conditional probability measures the probability of an event occurring when we already know that another event has occurred.

Formula
P(A|B) = P(A ∩ B) / P(B)

Where:

P(A|B) = Probability of A given B

P(A ∩ B) = Probability of both A and B

P(B) = Probability of B

# Step 2 – Demonstrate the Concept
Numerical Example
Suppose a class has:

20 students

12 students know Python.

8 of those 12 students also know SQL.

If we know that a student knows Python, the probability that the student also knows SQL is:

P(SQL | Python) = 8 / 12

= 0.667

So the conditional probability is approximately 66.7%.

# Real-World Example
A doctor may estimate the probability that a patient has a particular disease given that the patient has a specific test result.

# Business Example
An e-commerce company may calculate:

Probability of purchase given that a customer clicked an advertisement.

# AI/ML Example
Conditional probability is fundamental to classification and probabilistic machine-learning algorithms.

How does this demonstrate the concept?
The probability of SQL knowledge is calculated only among students who are already known to have Python knowledge.

# Step 3 – Implement the Concept

In [7]:

python_students = 25
python_and_sql_students = 10

conditional_probability = python_and_sql_students / python_students

print("P(SQL | Python):", conditional_probability)
print("Percentage:", conditional_probability * 100, "%")

P(SQL | Python): 0.4
Percentage: 40.0 %


# What Does the Program Do?
The program calculates the probability of knowing SQL given that the student already knows Python.

# What Statistical Concept Does It Demonstrate?
It demonstrates conditional probability.

# How Do We Interpret the Output?
The probability is approximately 0.667 or 66.7%.

This means that among the students who know Python, approximately 66.7% also know SQL.

# Where Is It Used in AI/ML?
Conditional probability is used in classification, Bayesian methods, recommendation systems, and probabilistic prediction.

# Bayes' Theorem
# Step 1 – Understand the Concept
# Definition
Bayes' Theorem is used to update the probability of an event when new evidence becomes available.

It combines:

Prior probability
New evidence
Likelihood of the evidence
Updated probability
Formula
P(A|B) = [P(B|A) × P(A)] / P(B)

Where:

P(A|B) = Posterior probability

P(B|A) = Likelihood

P(A) = Prior probability

P(B) = Probability of evidence

# Step 2 – Demonstrate the Concept
Numerical Example
Suppose:

1% of products from a factory are defective.

A quality-control test correctly identifies a defective product 90% of the time.

The test also incorrectly identifies a good product as defective 5% of the time.

Bayes' Theorem can be used to calculate the probability that a product is actually defective when the test says "defective."

The important idea is that the test result updates our original probability of defect.

# Real-World Example
Medical testing is a common application of Bayes' Theorem.

A test result can update the probability that a patient has a disease based on the initial prevalence and test accuracy.

# Business Example
A bank can use Bayesian reasoning to update the probability of loan default after receiving new customer information.

# AI/ML Example
Bayesian methods use prior knowledge and observed evidence to update predictions.

Bayes' Theorem is also fundamental to Naive Bayes classification.

# How does this demonstrate the concept?
Bayes' Theorem shows how new evidence changes an existing probability.

# Step 3 – Implement the Concept

In [8]:
p_defective = 0.01
p_test_defective_given_defective = 0.90
p_test_defective_given_good = 0.05

p_good = 1 - p_defective

p_test_defective = (
    p_test_defective_given_defective * p_defective
    + p_test_defective_given_good * p_good
)

posterior_probability = (
    p_test_defective_given_defective * p_defective
) / p_test_defective

print("Probability that the product is defective given a positive test:",
      posterior_probability)

print("Percentage:",
      posterior_probability * 100, "%")


Probability that the product is defective given a positive test: 0.15384615384615385
Percentage: 15.384615384615385 %


# What Does the Program Do?
The program applies Bayes' Theorem to calculate the probability that a product is actually defective after receiving a positive test result.

# What Statistical Concept Does It Demonstrate?
It demonstrates Bayesian probability and probability updating.

# How Do We Interpret the Output?
The result represents the updated probability of a product being defective after the test reports a defect.

This probability is different from the original 1% defect rate because new evidence has been introduced.

# Where Is It Used in AI/ML?
Bayesian reasoning is used in Naive Bayes classification, Bayesian networks, spam detection, medical diagnosis systems, and probabilistic inference.

# Probability Rules
# Step 1 – Understand the Concept
# Probability has several important rules.

Rule 1 – Probability Range
0 ≤ P(A) ≤ 1

Rule 2 – Complement Rule
P(Aᶜ) = 1 - P(A)

Rule 3 – Addition Rule
For mutually exclusive events:

P(A ∪ B) = P(A) + P(B)

For general events:

P(A ∪ B) = P(A) + P(B) - P(A ∩ B)

Rule 4 – Multiplication Rule
For independent events:

P(A ∩ B) = P(A) × P(B)

For dependent events:

P(A ∩ B) = P(A) × P(B|A)

# Step 2 – Demonstrate the Concept
Numerical Example
Suppose:

P(A) = 0.6

Using the complement rule:

P(Aᶜ) = 1 - 0.6

P(Aᶜ) = 0.4

Therefore, if the probability of a customer purchasing a product is 60%, the probability of the customer not purchasing it is 40%.

Real-World Example
A weather service may report a 70% probability of rain.

The probability of no rain is:

1 - 0.70 = 0.30

# Business Example
If a business estimates a 75% probability that a customer will renew a subscription, the probability of not renewing is 25%.

# AI/ML Example
Probability rules are used when combining class probabilities, calculating joint probabilities, and modeling uncertainty.

# How does this demonstrate the concept?
Probability rules allow us to calculate related events and combine probabilities correctly.

# Step 3 – Implement the Concept

In [9]:
p_purchase = 0.75

p_not_purchase = 1 - p_purchase

print("Probability of purchase:", p_purchase)
print("Probability of not purchasing:", p_not_purchase)

Probability of purchase: 0.75
Probability of not purchasing: 0.25


# What Does the Program Do?
The program uses the complement rule to calculate the probability of an event not occurring.

# What Statistical Concept Does It Demonstrate?
It demonstrates the complement rule.

# How Do We Interpret the Output?
If the probability of purchase is 0.75, the probability of not purchasing is 0.25.

Together they add up to 1.

# Where Is It Used in AI/ML?
Probability rules are used throughout probabilistic machine learning, classification, prediction, and uncertainty estimation.

# Random Variables
# Step 1 – Understand the Concept
# Definition
A random variable is a variable that assigns a numerical value to the outcome of a random experiment.

There are two main types:

Discrete Random Variable
A discrete random variable has countable values.

# Example:

Number of customers visiting a store.

Possible values:

0, 1, 2, 3, ...

Continuous Random Variable
A continuous random variable can take any value within a range.

# Example:

Height, weight, temperature, or delivery time.

# Step 2 – Demonstrate the Concept
Numerical Example
Suppose we roll a die.

Let X represent the number obtained.

Then:

X ∈ {1, 2, 3, 4, 5, 6}

X is a discrete random variable.

# Real-World Example
The number of customers entering a shop in one hour is a discrete random variable.

The exact waiting time of a customer is a continuous random variable.

# Business Example
A company can model the number of purchases made per hour as a discrete random variable.

It can model delivery time as a continuous random variable.

# AI/ML Example
Random variables are fundamental to probabilistic models and are used to represent uncertain quantities in machine learning.

# How does this demonstrate the concept?
The random variable converts outcomes of a random experiment into numerical values that can be analyzed mathematically.

# Step 3 – Implement the Concept

In [10]:

import numpy as np

dice_rolls = np.random.randint(1, 7, 10)

print("Random Dice Rolls:", dice_rolls)
print("Possible Values:", [1, 2, 3, 4, 5, 6])

Random Dice Rolls: [2 6 1 5 2 6 5 2 3 5]
Possible Values: [1, 2, 3, 4, 5, 6]


# What Does the Program Do?
The program simulates 10 rolls of a six-sided die.

# What Statistical Concept Does It Demonstrate?
It demonstrates a discrete random variable.

# How Do We Interpret the Output?
Each generated number represents the value of the random variable X for one dice roll.

The possible values are 1 through 6.

# `Where Is It Used in AI/ML?
Random variables are fundamental to probability distributions, probabilistic models, simulations, and machine-learning algorithms.

# Expected Value
# Step 1 – Understand the Concept
# Definition
Expected value is the long-run average outcome of a random variable when the experiment is repeated many times.

It is calculated by multiplying each possible outcome by its probability and then adding the results.

Formula
E(X) = Σ [x × P(X = x)]

Where:

x = Possible outcome

P(X = x) = Probability of that outcome

# Step 2 – Demonstrate the Concept
Numerical Example
Suppose a game has two possible outcomes:

Win = ₹100 with probability 0.4

Lose = ₹0 with probability 0.6

Expected Value:

E(X) = (100 × 0.4) + (0 × 0.6)

E(X) = 40

The expected value is ₹40.

This does not mean that every game will give exactly ₹40.

It means that over many repeated games, the average outcome approaches ₹40 under the stated probabilities.

# Real-World Example
An insurance company can use expected value to estimate the average expected claim cost based on different possible claim amounts and their probabilities.

# Business Example
A company can calculate the expected revenue from different business outcomes and their probabilities.

# AI/ML Example
Expected value is used in probabilistic prediction, decision-making under uncertainty, reinforcement learning, and statistical estimation.

# How does this demonstrate the concept?
Expected value combines possible outcomes with their probabilities to calculate the long-run average outcome.

# Step 3 – Implement the Concept

In [11]:
outcomes = np.array([100, 0])
probabilities = np.array([0.5, 0.6])

expected_value = np.sum(outcomes * probabilities)

print("Expected Value: ₹", expected_value)


Expected Value: ₹ 50.0


# What Does the Program Do?
The program multiplies each possible outcome by its probability and adds the results.

# What Statistical Concept Does It Demonstrate?
It demonstrates expected value.

# How Do We Interpret the Output?
The expected value is ₹40.

This represents the long-run average outcome if the same probabilistic situation is repeated many times.

# Where Is It Used in AI/ML?
Expected value is used in probabilistic prediction, decision-making, reinforcement learning, and statistical modeling.